File: cursor_expand/debug_expand/src/disasm/mn1613/labels.ts
/**
 * 逆アセンブラ用ラベル表（CDB / 手動ペア）
 * 根拠: emulater_code_test.mdc（CDB はバイトアドレス）
 */

import { parseCdb } from "../../load/cdb";
import type { Mn1613LabelPair } from "./types";

/**
 * アドレス → ラベル名の対応を持つ。
 * 同一アドレスに複数あるときはグローバル（G）を優先し、同じなら先勝ち。
 */
export class Mn1613LabelTable {
  private readonly byAddr = new Map<number, { name: string; scope: string }>();

  /**
   * CDB テキストからラベルを取り込む。
   * @param cdbText SDCC CDB（`L:` の末尾はバイトアドレス）
   */
  loadCdb(cdbText: string): void {
    const table = parseCdb(cdbText);
    for (const sym of table.symbols) {
      this._put(sym.wordAddr & 0xffff, sym.name, sym.scope);
    }
  }

  /**
   * ラベル:ワードアドレスの組を登録する（明示指定は既存を上書き）。
   * @param entries ペア列
   */
  setLabels(entries: Iterable<Mn1613LabelPair>): void {
    for (const e of entries) {
      const addr = e.wordAddr & 0xffff;
      const name = e.name.trim();
      if (!name) continue;
      this.byAddr.set(addr, { name, scope: "G" });
    }
  }

  /**
   * 1 件追加する。同一アドレスに G があるとき F/L は無視。
   * @param name ラベル名
   * @param wordAddr ワードアドレス
   * @param scope CDB スコープ。省略時は上書き
   */
  addLabel(name: string, wordAddr: number, scope = "G"): void {
    const n = name.trim();
    if (!n) return;
    this._put(wordAddr & 0xffff, n, scope);
  }

  /**
   * ワードアドレスに対応するラベル名を返す。
   * @param wordAddr ワードアドレス
   * @returns ラベル名。無ければ undefined
   */
  lookup(wordAddr: number): string | undefined {
    return this.byAddr.get(wordAddr & 0xffff)?.name;
  }

  /**
   * 登録件数。
   * @returns ユニークなアドレス数
   */
  get size(): number {
    return this.byAddr.size;
  }

  /**
   * アドレスへラベルを載せる。G は常に勝ち、F/L は空きまたは非 G のときだけ。
   * @param addr ワードアドレス
   * @param name ラベル名
   * @param scope G / F / L など
   */
  private _put(addr: number, name: string, scope: string): void {
    const prev = this.byAddr.get(addr);
    if (!prev || scope === "G" || prev.scope !== "G") {
      this.byAddr.set(addr, { name, scope });
    }
  }
}

File: retrocpu_test_framework_rs/src/framework/mn1613/disasm/labels.rs
//! 逆アセンブラ用ラベル表（CDB / 手動ペア）。
//!
//! 根拠: emulater_code_test.mdc（CDB はバイトアドレス）

use std::collections::HashMap;

use crate::framework::mn1613::cdb::parse_cdb;
use crate::error::FrameworkError;

/// ワードアドレスとラベル名のペア（初期化・手動登録用）。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Mn1613LabelPair {
    /// ラベル名
    pub name: String,
    /// ワードアドレス（16bit）
    pub word_addr: u16,
}

#[derive(Debug, Clone)]
struct LabelEntry {
    name: String,
    scope: String,
}

/// アドレス → ラベル名の対応を持つ。
///
/// 同一アドレスに複数あるときはグローバル（G）を優先し、同じなら先勝ち。
#[derive(Debug, Default)]
pub struct Mn1613LabelTable {
    by_addr: HashMap<u16, LabelEntry>,
}

impl Mn1613LabelTable {
    /// CDB テキストからラベルを取り込む（`L:` の末尾はバイトアドレス → ワード）。
    pub fn load_cdb(&mut self, cdb_text: &str) -> Result<(), FrameworkError> {
        let table = parse_cdb(cdb_text)?;
        for sym in &table.symbols {
            self.put((sym.word_addr & 0xffff) as u16, &sym.name, &sym.scope);
        }
        Ok(())
    }

    /// ラベル:ワードアドレスの組を登録する（明示指定は既存を上書き）。
    pub fn set_labels(&mut self, entries: impl IntoIterator<Item = Mn1613LabelPair>) {
        for e in entries {
            let name = e.name.trim();
            if name.is_empty() {
                continue;
            }
            self.by_addr.insert(
                e.word_addr & 0xffff,
                LabelEntry {
                    name: name.to_string(),
                    scope: "G".into(),
                },
            );
        }
    }

    /// 1 件追加する。同一アドレスに G があるとき F/L は無視。
    pub fn add_label(&mut self, name: &str, word_addr: u16, scope: &str) {
        let n = name.trim();
        if n.is_empty() {
            return;
        }
        self.put(word_addr & 0xffff, n, scope);
    }

    /// ワードアドレスに対応するラベル名を返す。
    pub fn lookup(&self, word_addr: u16) -> Option<&str> {
        self.by_addr
            .get(&(word_addr & 0xffff))
            .map(|e| e.name.as_str())
    }

    /// 登録件数（ユニークなアドレス数）。
    pub fn len(&self) -> usize {
        self.by_addr.len()
    }

    /// 登録が空か。
    pub fn is_empty(&self) -> bool {
        self.by_addr.is_empty()
    }

    fn put(&mut self, addr: u16, name: &str, scope: &str) {
        let prev = self.by_addr.get(&addr);
        if prev.is_none() || scope == "G" || prev.map(|p| p.scope.as_str()) != Some("G") {
            self.by_addr.insert(
                addr,
                LabelEntry {
                    name: name.to_string(),
                    scope: scope.to_string(),
                },
            );
        }
    }
}
